Add absolute millisecond tick count to RTOS classes

kjbracey-arm opened the pull request:
Give C++ access to the RTOS's absolute timebase, reducing the need to run private Timers and similar. Allows wait_until functionality, and makes it easier to avoid time drift.

Explicitly state that ticks are milliseconds in mbed OS, despite CMSIS-RTOS 2 permitting different tick rates.

Try to cover over the breaking API change potentially upcoming in CMSIS-RTOS 2.1.1,
when it reduces the tick count from 64-bit to 32-bit. (See https://github.com/ARM-software/CMSIS_5/issues/277)

Add "wait_until" methods to Thread, ConditionVariable, Semaphore and Mutex.

See also #3648 (wait_until for condition variables) and #5378 (EventQueue should use RTOS tick count).

kjbracey-arm commented:
This is a straw man to help kick https://github.com/ARMmbed/mbed-os/pull/3648 along - I really want that to have `wait_until`, and lack of C++ API to get current time has been viewed as a blocker for that.

This would be a minimal addition to the C++ API to make `ConditionVariable::wait_until` useable with only C++ APIs.

But do we really need every CMSIS-RTOS API in the C++ API? Is the C++ API supposed to be a complete abstraction layer providing all facilities of the CMSIS-RTOS abstraction layer? Is there anything wrong with expecting application to use osKernelGetTickCount?

@c1728p9, @pan-, @0xc0170, @geky, @sg-

kjbracey-arm commented:
(As an aside, I'm was surprised to see the CMSIS-RTOS 2 tick count being 64-bit. The API documentation all says it's 32-bit).

0xc0170 commented:
> But do we really need every CMSIS-RTOS API in the C++ API? Is the C++ API supposed to be a complete abstraction layer providing all facilities of the CMSIS-RTOS abstraction layer? Is there anything wrong with expecting application to use osKernelGetTickCount directly?

User space facing C++ API always, providing us also keeping it backward compatible (cmsis API is evolving, changing).

Is the proper placement for these 2 new functions (get tick count and freq) as part of Thread ? However as RTOSTimer is deprecated, I do not see any other class that would fit better apart from having a new one. Just those 2 attributes are not related to a Thread.  New `Kernel` class to handle these?

kjbracey-arm commented:
> User space facing C++ API always, providing us also keeping it backward compatible (cmsis API is evolving, changing).

Sounds like CMSIS-RTOS is failing in its purpose then :(

> However as RTOSTimer is deprecated, I do not see any other class that would fit better apart from having a new one. Just those 2 attributes are not related to a Thread. New Kernel class to handle these?

I don't disagree - my first thought was `Kernel::get_tick_count`, but thought I'd try to get away without a whole new class.

A lot of the static Thread functions are debatable. The existing `Thread::wait()` notably doesn't map to a CMSIS-RTOS `osThread` function. I was kind of trying to segue from `wait()` to `wait_until()` to `get_tick_count()` to support wait_until with no-one noticing.

I guess if we think we'll want more `osKernel` functions...

geky commented:
I've found it weird for a while that we already have some global timing functions (wait_ms, sleep) but no function for access to some sense of global time.

RTOS aside, why don't we just provide a high-level tick function (hooked to rtos when possible)?

``` c++
float mbed_tick();
mbed_tick_t mbed_tick_us();
mbed_tick_t mbed_tick_ms();
```

It's a common API, I find it strange it's missing in mbed OS.

c1728p9 commented:
All of the user facing APIs use seconds/ms/us as a timebase and this is unlikely to change. **We should seriously consider fixing the rtos tickrate to 1ms** align with this. This won't add substantially to power consumption since with tickless RTX a tick will only occur when threads are already running. When all the threads are suspend the OS will stop ticking entirely. If this is done then the addition of the wait_until APIs will be greatly simplified and easier for our users.

As for the global timing I agree that we definitely need these functions. Some kind of system wide timer which is available for both mbed 2 (backed by ticker or lpticker) and mbed 5 (backed by os) would be nice.

Some thoughts about possible C/C++ API for this:
```
SystemTime::wait(time_ms);
SystemTime::wait_until(time_ms);
SystemTime::read_ms();
mbed_sys_wait(time_ms);
mbed_sys_wait_until(time_ms);
mbed_sys_read_ms();
```

kjbracey-arm commented:
I agree with Russ that just defining "RTOS ticks == milliseconds" would avoid a lot of grief. Tick frequency really doesn't matter much, so doing that gains simplicity at basically no cost. Nothing wrong with choosing a value and sticking with it.

Yes, would be useful to have the same thing available for mbed OS 2. Presumably there's an overhead issue there, so you wouldn't initialise it until first read? In effect it would be an on-demand global ticker, a bit like the on-demand global event queue? 

For mbed OS 5, it needs to be defined that this API is representing the RTOS ticks, thus is phase aligned with all the milliseconds in all the RTOS timeouts.

@geky - I don't see how `float mbed_tick()` works - gets less accurate the longer you're up? Or wraps at some point? And `mbed_tick_us()` implies a high-res high-power timer running forever after the first call to it, so not the sort of thing to be called lightly. I think people can probably access the us_ticker APIs on demand if they really want that precision. I'd restrict this to the milliseconds, thus totally free on mbed OS 5, and only costing you an lp_timer on mbed OS 2.

I think the "mbed_time_api.h" or whatever, with RTOS + non-RTOS versions can follow - figuring out exactly what the implementation is for non-RTOS is more complex. If we assume it calls the C++ RTOS API for the RTOS case, then we can pin that RTOS API down first.

0xc0170 commented:
Great to see some alignment here. Having milliseconds , available for both mbed 2 and mbed OS.

> As for the global timing I agree that we definitely need these functions. Some kind of system wide timer which is available for both mbed 2 (backed by ticker or lpticker) and mbed 5 (backed by os) would be nice.

👍  Users just used HAL us ticker getter functions to get time, this would finally replace this and have something "steady" :-) How it should look like?

Just a note: long time ago we considered that once we can add chrono-like API.  We are not yet c++11, worht mentioning.

What I would consider is something similar to above `SystemTime`. Should all be unified to one or having all available - we are having already in the system:

-  `high_resolution_clock` (us ticker) - not steady
- `steady_clock` (kernel ticker - running on lp ticker?)
- `system_clock` (this could be rtc or steady clock?) - not certain about his, might be just first two that is sufficient for a user

Here some thoughts (pseudocode warning):

This would be C++ API used, new class introduction:

```
// just nested classes here, I would prefer namespacing but we do not use that much
class SystemTick {
    // on demand clock, need to be init and freed, would be nice?!
    class high_resolution_clock {
        public:
            using time_point = timestamp_t;

            static time_point  now() {
                // or invokes Timer classes methods
                return ticker_read(us_ticker_data);
            }

            static bool is_steady {
                return true;
            }
    }

    class steady_clock {
        public:
            using time_point = uint64_t;

            static time_point now() {
                return osKernelTickerCount();
            }

            static bool is_steady {
                return true;
            }
    }
}

```

what types do we have here? high resolution clock - timestamp_t that is 64bit wide. steady_clock - using ms resolution, 64bit also possibly? 

This can be extended for instance:

```
    class realtime_clock {
            using time_point = time_t;

            static time_point now() {
                return rtc_read();
            }

            static bool is_steady {
                return false;
            }
    }
```


I might like if high resolution clock is on demand, thus needs created on demand and destroyed if not used (ctor/dtor would take care of it) - this is not what I have seen in chrono lib, it is always accessible. stead_clock or realtime_clock are always running, thus do not need this?

Usage:

```
{
    SystemTick::high_resolution_clock();
    time_point now = SystemTick::high_resolution_clock.now();
    // do something
    time_point end = SystemTick::high_resolution_clock.now();
    if (end - now > 10) {
        // etc
    }
}
// high_resolution_clock is off here

```

This would be for mbed OS. The only diff above would be for steady_clock(), as RTOS is not present there, they would as noted above by @c1728p9 map to us ticker for instance. I used `now` but we have in our codebase `read()` thus this could be changed for consistency.

`Wait()` and `wait_until`, should they be part of SystemTick?

cc @pan- - might have some thoughts

kjbracey-arm commented:
Quick rework, into having a `Kernel` class, including the "ticks == milliseconds" statement.

By re-orienting into being a Kernel API, I'm kind of trying to side-step the bigger discussion about what time APIs and timebases would be needed independent of RTOS, and lock down access to the specific things that the RTOS already provides - in particular, reading the tick count that matches what we're passing to all the semaphore waits etc.

A wider chrono-style API, or any other sort of RTOS-independent time API, could layer one of its clocks on top of this in the RTOS case.

As it stands, this would unblock `ConditionVariable::wait_until`, which is my particular interest.

pan- commented on rtos/Kernel.h:
```diff
@@ -0,0 +1,98 @@
+ *
+ * The Kernel object cannot be instantiated - all functions are static.
+ */
+class Kernel {
```
Why is this a class instead of a namespace ?

kjbracey-arm replied:
No good answer to that. I don't have a preference either way.

My aim was merely that these methods appear much the same as all the static `Thread` methods to casual users. A namespace might be a nicer way of doing that.

pan- replied:
In practice it doesn't change much; namespace or class with only static members are not very different however semantically a namespace indicates that `kernel` is **not** an object and therefore cannot be instantiated as one; using a namespace clarifies the intent. 

It also avoid to reference Kernel as a data structure in the documentation generated by doxygen.

c1728p9 commented on rtos/Kernel.cpp:
```diff
@@ -0,0 +1,52 @@
+namespace rtos {
+
+uint64_t Kernel::get_ms_count() {
+    return osKernelGetTickCount();
```
I think this was changed back to uint32_t in more recent versions of RTX:
https://github.com/ARM-software/CMSIS_5/commit/9470b0bfa35625542f54afb99ec8ddf70ea8f756#diff-818bf32b54ffc78cfbdfe02740cf27f2R343

kjbracey-arm replied:
Gah. Okay, that's going to cause grief. Are we moving to 2.1.1 in the immediate future?

I want this API to be stable - I guess we'd make it 32-bit now, and make sure it's written to cope with both 32-bit and 64-bit underlying?

Also, only 2.1.1 says it's okay to call from ISR, 2.1.0 says it isn't.

kjbracey-arm replied:
Brother. Looking deeper, RTX 5.1.1 does actually permit call from interrupt, so could let that pass.

But looking at RTX 5.2.0, I think the implementation (and possibly the CMSIS 2.1.1 spec) is flakey. Wrapping isn't handled sensibly. If the current time is 0x8000 and you say `delay_until(0x7ffe)`, it assumes you want to wait until 2^32-2 ticks in the future, not that you've overrun and are 2 ticks late. That's not a sensible assumption - it would be more conventional to assume half the number space is past and half is future.

2.1.0 is solid, because it returns immediately in that case - it does 64-bit maths, computes "target-now" as 2^64-2, and decides that's not a sensible wait because it is >2^31, and returns osErrorParameter immediately. (Arguably it maybe could have said OK, but it does return immediately, which is the main thing).

I would be very wary about actually using this API with 2.1.1 behaviour. Get CPU starved a bit too long and you're sleeping for 49 days.

0xc0170 replied:
> Gah. Okay, that's going to cause grief. Are we moving to 2.1.1 in the immediate future?

I would assume we would. There are couple of changes incoming (cortex-a, cortex m33) that might require some updates (some bugs were found and are still being in discussion).

@JonatanAntoni are our findings correct that are described above?

JonatanAntoni replied:
Yes, we decided to go back to uint32_t as tick type with CMSIS 5.1 release. osDelayUntil will wait exactly until the given tick time is reached (again). So if you call osDelayUntil with a value slightly lower than osKernelGetTickCount it'll wait 49 days, sure. I think we don't have a chance to predict if you really want to wait 49 days or not. Please let us know if you have reasonable concerns about the current implementation.

kjbracey-arm replied:
Just raised an issue over here, going in to detail on my implementation concern: https://github.com/ARM-software/CMSIS_5/issues/277

If I fix this API at 32-bit, and we need to support 2.1.0, I'm going to have to muck around anyway with 32-bit/64-bit adaptation, so it's not going to be straight mapped to osDelayUntil, so maybe I could also muck around in 2.1.1 to adjust the past handling to "31-bit" to suit my taste...

As it happens, I'm not intending to use osDelayUntil at this point anyway, just have an analogue to it in ConditionVariable::wait_until() - its timeout will be based on Semaphore::Wait, so it is free to have its own absolute timestamp interpretation, but behaviour really should be aligned.

kjbracey-arm replied:
On the 32-bit/64-bit difference - how about making the timestamp be a typedef, so it matches and adapts to the underlying system? Would be less problematic than trying to handle 32<->64 bit conversions.

Normal users shouldn't care whether it's just 32 or 64 bit, just that the "get ticks" and "until" APIs use the same numbers.

Although I'm not sure how I'd detect it, without grobbling into RTX headers - it's a compile-time difference and there doesn't seem to be a compile-time version macro in "cmsis_os2.h". And we're not using C++11, so no `decltype(osKernelGetTickCount())`.

c1728p9 commented on rtos/Kernel.cpp:
```diff
@@ -0,0 +1,52 @@
+    return osDelayUntil(millisec);
+}
+
+int32_t Kernel::lock() {
```
Do you have any use cases for lock and unlock in mind? I would be inclined to leave this private unless it is needed for a certain reason.

kjbracey-arm replied:
No immediate use, but it seemed like if adding a Kernel class, it was the most obvious API from osKernel that you people might want. Otherwise it will just be the tick count, if I remove the delays, as I'm now minded to.

c1728p9 commented on rtos/Thread.h:
```diff
@@ -317,13 +317,28 @@ class Thread : private mbed::NonCopyable<Thread> {
+      @param   millisec absolute time in millisec
+      @return  status code that indicates the execution status of the function.
+      @note not callable from interrupt
+      @note if the time in the past, or >= 2^32-1 ms in the future, this
```
This will need to be updated if the tick API changes to 32 bit in RTX (as mentioned above).

c1728p9 commented on rtos/Thread.h:
```diff
@@ -317,13 +317,28 @@ class Thread : private mbed::NonCopyable<Thread> {
+            returns with an immediate error
+      @note this call is equivalent to Kernel::delay_until()
+    */
+    static osStatus wait_until(uint64_t millisec);
```
What are you thoughts on dropping the Kernel class and adding get_ms_count() to `Thread` since the other timing functions are already here? The Kernel class could always be added later if kernel specific functionality is needed. This would also keep the API expansion to a minimum.

kjbracey-arm replied:
That was what my first draft looked like, but changed to this after @0xc0170's comments, which I tended to agree with.

I'm struggling a bit to find logic in CMSIS-RTOS's placement of stuff in osThread versus osKernel versus osDelay - why not "osThreadLock" and "osKernelDelay" and "osYield"?

Although, thinking harder I guess delay is specifically current thread only, so actually delay probably /shouldn't/ be in Kernel - delaying one thread doesn't affect anything else in the system, unlike the lock. I'm now inclined to take those out of Kernel.

pan- commented:
> What about keeping things simple and adding these two extra timing related functions just to the Thread class?

It make sense to have `delay` and `delay_until` in the `Thread` class because they apply to the caller thread however does it make sense for `get_ms_count` ? This function is not related to threading by any mean. Is it really simpler for the user to have APIs organized in an incoherent fashion ?

kjbracey-arm commented:
Well, this got complex fast. Added code to cope with a 32-bit counter from the RTOS and convert it to 64 bits, so we retain a lovely non-wrapping 64-bit count in case we move to API v 2.1.1, and avoid its "wait for 49 days" problem.

Also removed the delay functions from Kernel, as they don't act on the kernel as a whole.

c1728p9 commented on rtos/Thread.cpp:
```diff
@@ -334,6 +334,43 @@ osStatus Thread::wait(uint32_t millisec) {
     return osDelay(millisec);
 }
 
+osStatus Thread::wait_until(uint64_t millisec) {
```
One possible alternative, is rather than converting up to 64 bits, you could instead convert down to 32 bits. You could still handle the case there the time is slightly in the past inside wait_until. The code to perform the delay would also be simplified substantially.

This would make the behavior for things like Semaphores, Mutexes and Events implementing wait_until more consistent, since it would just be a single call, rather than a loop which will break any ordering RTX tries to keep.

sg- commented on rtos/Thread.h:
```diff
@@ -317,13 +317,31 @@ class Thread : private mbed::NonCopyable<Thread> {
+      @note if millisec indicates the current tick count it will return immediately,
+            either with "osOK" or an error.
+      @note depending on implementation, the maximum wait time may be limited
+            to 0xfffffffe milliseconds, which means specifying a time further
```
osWaitForever would match existing defaults and be sensible.

sg- commented on rtos/Kernel.h:
```diff
@@ -0,0 +1,77 @@
+     @return  negative osStatus error code on error
+     @note not callable from interrupt
+ */
+int32_t lock();
```
This seems like a bad idea to expose as an API.

sg- commented on rtos/Kernel.h:
```diff
@@ -0,0 +1,77 @@
+     @return  negative osStatus error code on error
+     @note not callable from interrupt
+ */
+int32_t unlock();
```
This seems like a bad idea to expose as an API.

kjbracey-arm commented on rtos/Thread.cpp:
```diff
@@ -334,6 +334,43 @@ osStatus Thread::wait(uint32_t millisec) {
+        do {
+            // RTX internals seem to avoid 0xFFFFFFFF delays, so do the same, even
+            // though nothing in documentation says osDelay(0xFFFFFFFFF) shouldn't work.
+            const uint32_t max_delay = 0xFFFFFFFE;
```
This isn't part of the API, it's just caution on the loop - "how big a chunk will we break this into for the underlying osDelay call?" The number is effectively arbitrary. I'll scrub the loop anyway.

kjbracey-arm commented:
Rebased now that ConditionVariable exists, and go through adding "until" calls to it and other classes.

All "wait_untils" have the simplifying assumption that permits them to return early if the delay is excessive (specified as >24 days, actually >49 days in current implementation).

pan- commented on rtos/Mutex.cpp:
```diff
@@ -58,11 +59,30 @@ osStatus Mutex::lock(uint32_t millisec) {
+
+bool Mutex::trylock_for(uint32_t millisec) {
+    osStatus status = lock(millisec);
+    if (status == osOK) {
```
I suppose _count should be incremented in case of success.

kjbracey-arm replied:
`lock(millisec)` already does that - this is just munging the return value (and asserting there's no programming error).

0xc0170 commented:
@c1728p9 @sg- Could you please review the latest changes?

c1728p9 commented on rtos/Thread.cpp:
```diff
@@ -334,6 +334,43 @@ osStatus Thread::wait(uint32_t millisec) {
+    // by looking at the return type of osKernelGetTickCount. We assume
+    // our header at least matches the implementation, so we don't try looking
+    // at the run-time version report. (There's no compile-time version report)
+    if (sizeof(osKernelGetTickCount() == sizeof(uint64_t))) {
```
Shouldn't this be
```
sizeof(osKernelGetTickCount()) == sizeof(uint64_t)
```

rather than
```
sizeof(osKernelGetTickCount() == sizeof(uint64_t))
```

kjbracey-arm replied:
Whoops! Good spot. Totally failed to grasp what you were saying earlier...

c1728p9 commented:
It looks good. I just have one question about the use of `sizeof` - https://github.com/ARMmbed/mbed-os/pull/5419/files#r151728337

kjbracey-arm commented:
Corrected the sizeof bug @c1728p9 pointed out, and rebased.

Had to patch up minor conflicts with the recent `ConditionVariable` list fix, no other changes.

0xc0170 commented:
@kjbracey-arm Please review jekins CI failures, look related

Once that CI is green, we can run morph CI.

This awaits @sg- approval, currently it's requested changes.

kjbracey-arm commented:
Cleared the Jenkins error.

0xc0170 commented:
/morph build

kegilbert commented:
Test machine crashed, rekicking off
/morph test

geky commented:
So are we getting a `what_time_is_it_in_milliseconds_mod_2_32()` function that doesn't rely on an RTOS call?

kjbracey-arm commented:
I think we should have such a thing - something that like `::wait_ms()` routes to either this RTOS version or something using a Timer/whatever.

I don't feel familiar enough with all the HAL-based timer stuff to know how what the most efficient/correct/portable non-RTOS implementation is though, so not sure if I'm the best person to tackle it.

Since earlier discussion, I've noticed that there is a C library `clock()` implementation in mbed_retarget.cpp. Could that be massaged into shape, rather than adding a new custom API? But perhaps there are problems making sure clock_t and CLOCKS_PER_SEC are what we need across the various C libraries...

adbridge commented:
@sg- This is still pending your approval to be progressed..

kjbracey-arm commented:
Rebased - added "ISR or not" comments to my new calls to match changes made to other calls.

Added one more commit that corrects all existing `ConditionVariable` methods comments - it's not at all useable from ISR.

cmonr commented:
/morph build

cmonr commented:
@sg-, can you confirm that your comments were addressed?

cmonr commented:
/morph build

kjbracey-arm commented:
Rebased as I hear it may solve LP ticker test issues.

cmonr commented:
/morph build

cmonr commented:
@sg- Are you alright with the updated commits?

cmonr commented:
Waiting on @sg-. If no response, will merge EoD Monday (CST).
